PINCache/PINOperationQueue.m

```diff
@@ -0,0 +1,217 @@
+
++ (instancetype)operationWithBlock:(dispatch_block_t)block reference:(id <PINOperationReference>)reference;
+
+@end
+
+@implementation PINOperation
+
++ (instancetype)operationWithBlock:(dispatch_block_t)block reference:(id<PINOperationReference>)reference
+{
+  PINOperation *operation = [[self alloc] init];
+  operation.block = block;
+  operation.reference = reference;
+
+  return operation;
+}
+
+@end
+
+@implementation PINOperationQueue
+
+- (instancetype)initWithMaxConcurrentOperations:(NSUInteger)maxConcurrentOperations
+{
+  if (self = [super init]) {
+    NSAssert(maxConcurrentOperations > 1, @"Max concurrent operations must be greater than 1. If it's one, just use a serial queue!");
+    _operationReferenceCount = 0;
+    
+    pthread_mutexattr_t attr;
+    pthread_mutexattr_init(&attr);
+    pthread_mutexattr_settype(&attr, PTHREAD_MUTEX_RECURSIVE);
+    pthread_mutex_init(&_lock, &attr);
+    
+    _serialQueue = dispatch_queue_create("PINOperationQueue Serial Queue", DISPATCH_QUEUE_SERIAL);
+    
+    _concurrentQueue = dispatch_queue_create("PINOperationQueue Unprioritized Serial Queue", DISPATCH_QUEUE_CONCURRENT);
+    _concurrentSemaphore = dispatch_semaphore_create(maxConcurrentOperations - 1);
+    _semaphoreQueue = dispatch_queue_create("PINOperationQueue Serial Semaphore Queue", DISPATCH_QUEUE_SERIAL);
+    
+    _allOperations = [[NSMutableOrderedSet alloc] init];
+    _lowPriorityOperations = [[NSMutableOrderedSet alloc] init];
+    _defaultPriorityOperations = [[NSMutableOrderedSet alloc] init];
+    _highPriorityOperations = [[NSMutableOrderedSet alloc] init];
+  }
+  return self;
+}
+
+- (void)dealloc
+{
+  pthread_mutex_destroy(&_lock);
+}
+
+- (id <PINOperationReference>)nextOperationReference
+{
+  [self lock];
+    id <PINOperationReference> reference = [NSNumber numberWithUnsignedInteger:++_operationReferenceCount];
+  [self unlock];
+  return reference;
+}
+
+- (id <PINOperationReference>)addOperation:(dispatch_block_t)block withPriority:(PINOperationQueuePriority)priority
+{
+  id <PINOperationReference> reference = [self nextOperationReference];
+  
+  NSMutableOrderedSet *queue;
+  switch (priority) {
+    case PINOperationQueuePriorityLow:
+      queue = _lowPriorityOperations;
+      break;
+      
+    case PINOperationQueuePriorityDefault:
+      queue = _defaultPriorityOperations;
+      break;
+      
+    case PINOperationQueuePriorityHigh:
+      queue = _highPriorityOperations;
+  }
+  
+  PINOperation *operation = [PINOperation operationWithBlock:block reference:reference];
+  
+  [self lock];
+    [queue addObject:operation];
+    [_allOperations addObject:operation];
+  [self unlock];
+  
+  [self scheduleOperations:NO];
+  
+  return reference;
+}
+
+- (void)cancelOperation:(id <PINOperationReference>)operationReference
+{
+  [self lock];
+    [_canceledOperations addObject:operationReference];
+  [self unlock];
+}
+
+- (void)scheduleOperations:(BOOL)onlyCheckSerial
```

Comment:
> this will actually execute maxConcurrentOperations + 1?

I think @garrettmoon covers this when he creates the semaphore:
`_concurrentSemaphore = dispatch_semaphore_create(maxConcurrentOperations - 1);`